### ISSUE TITLE
reversed outlines when building fonts from quadratic curves

### ISSUE THREAD

**Issue by RosaWagner** [googlefonts/fontmake#1045]
We are trying to onboard a font with quadratic sources and manual hinting. The problem is that the outlines get reversed during export (as if the sources were postscript) and the manual hinting is then completely wrong. 

The fonts can be build with manual hinting with gftools builder in that branch: https://github.com/RosaWagner/intel-one-mono/tree/googlefonts

This is the image the designers send after testing: 

**Their build:**
![Intel One Mono hinting (FJT build 1_3)](https://github.com/googlefonts/fontmake/assets/12222436/42fe0aca-86e5-48b4-9354-e7a47eba7aee)

**Fontmake build:**
![Intel One Mono hinting (Google build)](https://github.com/googlefonts/fontmake/assets/12222436/a4626137-7fcd-482e-b6ae-207d6d1ffb1e)

cc @simoncozens @m4rc1e

**Comment by RosaWagner** [comment]
Ah it might not be supported by gftools builder

**Comment by anthrotype** [comment]
for a font with quadratic sources and manual hinting, in addition to --keep-direction, you could also pass `--ttf-curves=keep-quad` so cu2qu is skipped as well (it should not modify the quads but it's quicker if you know in advance there are no cubic to convert). Also --keep-overlaps if you're building a static font, so the original quadratic outlines are not messed with.

**Comment by anthrotype** [comment]
fontmake's defaults are geared towards the most common use case i.e. source containing cubic bezier (hence drawn following PostScript counter-clockwise direction) => TrueType with quadratic curves and clockwise direction

**Comment by RosaWagner** [comment]
> for a font with quadratic sources and manual hinting, in addition to --keep-direction, you could also pass --ttf-curves=keep-quad so cu2qu is skipped as well (it should not modify the quads but it's quicker if you know in advance there are no cubic to convert). Also --keep-overlaps if you're building a static font, so the original quadratic outlines are not messed with.

Okay thanks!

> fontmake's defaults are geared towards the most common use case i.e. source containing cubic bezier (hence drawn following PostScript counter-clockwise direction) => TrueType with quadratic curves and clockwise direction

yeah I am aware.

**Comment by simoncozens** [comment]
I feel like you shouldn't need to pass an optional flag to do the right thing. The right thing should be the default.

**Comment by anthrotype** [comment]
but how?

**Comment by anthrotype** [comment]
perhaps are you suggesting something like, if a glyph contours only contain quadratic curves (or any?) then fontmake can assume they are already in the correct direction and not flip it, whereas if there's all (or any?) cubic curves then proceed with the reversal?

**Comment by simoncozens** [comment]
Exactly.

**Comment by anthrotype** [comment]
or you want to sum the signed area of all the glyph's contours to obtain an overall direction and flip accordingly to match the recommeded orientation for a given output format?
that seems a bit risky..

**Comment by anthrotype** [comment]
yeah, there is a recommended order but fontmake isn't actually checking which order it currently is and acting based on that. fontmake doesn't know nor checks whether the current order is right or wrong; it can only know whether the glyph contains all-quadratic, all cubic, or a mix of the two which is also legal in sources.

**Comment by anthrotype** [comment]
technically a font could contain only lines, no curves at all, how do you know in that case what the right thing to do is?..
I'm very tempted to not change anything and keep it like this tbh.
